File: Template/String/Hash(double random mod).cpp
constexpr int N=100010;

bool isprime(i64 x){
    if(x<2)
        return false;
    for(int i=2;i<=x/i;i++){
        if(x%i==0)
            return false;
    }
    return true;
}

i64 findprime(i64 x){
    while(!isprime(x))
        x++;
    return x;
}

mt19937 rng(chrono::steady_clock::now().time_since_epoch().count());
const i64 Base=131ll;
const i64 mod1=findprime(rng()%900000000+100000000);
const i64 mod2=findprime(rng()%900000000+100000000);

i64 Hash1[N],Hash2[N];
i64 p1[N],p2[N];


pair<i64,i64> get(int l,int r){
	i64 h1=(Hash1[r]-Hash1[l-1]*p1[r-l+1]%mod1+mod1)%mod1;
	i64 h2=(Hash2[r]-Hash2[l-1]*p2[r-l+1]%mod2+mod2)%mod2;
	return {h1,h2};
}

int main(){
	ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n=10;
    string s=" abcdefghij";

    p1[0]=p2[0]=1;
    for(int i=1;i<=n;i++){
    	p1[i]=p1[i-1]*Base%mod1;
    	p2[i]=p2[i-1]*Base%mod2;
    }

    for(int i=1;i<=n;i++){
    	Hash1[i]=(Hash1[i-1]*Base%mod1+s[i]-'a'+1)%mod1;
    	Hash2[i]=(Hash2[i-1]*Base%mod2+s[i]-'a'+1)%mod2;
    }

    return 0;
}

File: Template/String/Hash(double random mod).py
class StringHash:
    def __init__(self,s:str):
        self.base=int(1e18)+3
        self.mod=randint(int(8e17),int(9e17))
        n=len(s)
        self.POW=[0]*(n+1)
        self.POW[0]=1
        for i in range(1,n+1):
            self.POW[i]=self.POW[i-1]*self.base%self.mod
        self.preffix=[0]*n
        self.preffix[0]=ord(s[0])
        for i in range(1,n):
            self.preffix[i]=self.preffix[i-1]*self.base+ord(s[i])
            self.preffix[i]%=self.mod

    def get_pre_hash(self,l,r)->int:
        return (self.preffix[r]-(self.preffix[l-1] if l else 0)*self.POW[r-l+1]%self.mod)%self.mod

    def is_same(self,l1,r1,l2,r2)->bool:
        return self.get_pre_hash(l1,r1)==self.get_pre_hash(l2,r2)

    def get_any_s_hash(self,s:str)->int:
    	n=len(s)
    	h=0
    	for i in range(n):
    		h=h*self.base+ord(s[i])
    		h%=self.mod
    	return h


#---------------------------------------


from random import randint


def isprime(x):
	if x<2:
		return False
	i=2
	while i<=x//i:
		if x%i==0:
			return False
		i+=1
	return True

def findprime(x):
	while not isprime(x):
		x+=1
	return x


n=10
s=' abcdefghij'

mod1=findprime(randint(1000000000,10000000000))
mod2=findprime(randint(1000000000,10000000000))
Base=131

h1=[0]*(n+1);h2=[0]*(n+1)
for i in range(1,n+1):
	h1[i]=(h1[i-1]*Base+ord(s[i])-ord('a')+1)%mod1
	h2[i]=(h2[i-1]*Base+ord(s[i])-ord('a')+1)%mod2


p1=[0]*(n+1);p2=[0]*(n+1)
p1[0]=p2[0]=1
for i in range(1,n+1):
	p1[i]=p1[i-1]*Base%mod1
	p2[i]=p2[i-1]*Base%mod2

def get(l,r):
	return [(h1[r]-h1[l-1]*p1[r-l+1])%mod1,(h2[r]-h2[l-1]*p2[r-l+1])%mod2]